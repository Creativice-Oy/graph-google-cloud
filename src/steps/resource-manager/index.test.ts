import {
  Recording,
  createMockStepExecutionContext,
  MockIntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-testing';
import { setupGoogleCloudRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../types';
import { fetchResourceManagerIamPolicy } from '.';
import { integrationConfig } from '../../../test/config';
import { iamSteps, IAM_USER_ENTITY_TYPE } from '../iam';

async function executeIamSteps(
  context: MockIntegrationStepExecutionContext<IntegrationConfig>,
) {
  for (const step of iamSteps) {
    await step.executionHandler(context);
  }
}

describe('#fetchResourceManagerIamPolicy', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchResourceManagerIamPolicy',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await executeIamSteps(context);
    await fetchResourceManagerIamPolicy(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const userEntities = context.jobState.collectedEntities.filter(
      (e) => e._type === IAM_USER_ENTITY_TYPE,
    );
    const iamServiceAccountEntities = context.jobState.collectedEntities.filter(
      (e) => e._type === 'google_cloud_iam_service_account',
    );
    const iamServiceAccountKeyEntities = context.jobState.collectedEntities.filter(
      (e) => e._type === 'google_cloud_iam_service_account_key',
    );
    const iamServiceAccountImplementsRoleRelationships = context.jobState.collectedRelationships.filter(
      (r) => r._type === 'google_cloud_iam_service_account_implements_role',
    );
    const iamServiceAccountHasKeyRelationships = context.jobState.collectedRelationships.filter(
      (r) => r._type === 'google_cloud_iam_service_account_has_key',
    );

    expect(userEntities.length).toBeGreaterThanOrEqual(1);
    expect(iamServiceAccountEntities.length).toBeGreaterThanOrEqual(1);
    expect(iamServiceAccountKeyEntities.length).toBeGreaterThanOrEqual(1);
    expect(
      iamServiceAccountImplementsRoleRelationships.length,
    ).toBeGreaterThanOrEqual(1);
    expect(iamServiceAccountHasKeyRelationships.length).toBeGreaterThanOrEqual(
      1,
    );

    expect(userEntities).toMatchGraphObjectSchema({
      _class: ['User'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_iam_user' },
          type: { type: 'string' },
          deleted: { type: 'boolean' },
          uniqueid: { type: 'string' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
        },
      },
    });

    expect(iamServiceAccountEntities).toMatchGraphObjectSchema({
      _class: ['User'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_iam_service_account' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          enabled: { type: 'boolean' },
          projectId: { type: 'string' },
          id: { type: 'string' },
          oauth2ClientId: { type: 'string' },
          etag: { type: 'string' },
        },
      },
    });

    expect(iamServiceAccountKeyEntities).toMatchGraphObjectSchema({
      _class: ['AccessKey'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_iam_service_account_key' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          origin: { type: 'string' },
          type: { type: 'string' },
          algorithm: { type: 'string' },
        },
      },
    });

    expect(iamServiceAccountImplementsRoleRelationships).toEqual(
      iamServiceAccountImplementsRoleRelationships.map((r) =>
        expect.objectContaining({
          _class: 'IMPLEMENTS',
        }),
      ),
    );

    expect(iamServiceAccountHasKeyRelationships).toEqual(
      iamServiceAccountHasKeyRelationships.map((r) =>
        expect.objectContaining({
          _class: 'HAS',
        }),
      ),
    );
  });
});
