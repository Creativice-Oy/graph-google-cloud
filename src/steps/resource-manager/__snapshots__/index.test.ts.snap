// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildOrgFolderProjectMappedRelationships should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "CzUfI2XTWvMJgVGlic9LWA==",
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2021-05-31T15:15:58.148Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "updatedOn": 1622474158148,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
    Object {
      "_class": Array [
        "Organization",
      ],
      "_key": "organizations/958457776463",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-26T21:15:56.407Z",
            "directoryCustomerId": "C048tgq5f",
            "displayName": "jupiterone.dev",
            "etag": "ZmYW46DY9wJtowGQ20pa3g==",
            "name": "organizations/958457776463",
            "state": "ACTIVE",
            "updateTime": "2021-05-26T21:15:57.254Z",
          },
        },
      ],
      "_type": "google_cloud_organization",
      "createdOn": 1622063756407,
      "directoryCustomerId": "C048tgq5f",
      "displayName": "jupiterone.dev",
      "etag": "ZmYW46DY9wJtowGQ20pa3g==",
      "lifecycleState": "ACTIVE",
      "name": "organizations/958457776463",
      "updatedOn": 1622063757254,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/299817291758",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-06-03T16:44:28.216Z",
            "displayName": "example-folder-level-0",
            "etag": "2f8zjBmyEStHXN45HiqtvA==",
            "name": "folders/299817291758",
            "parent": "organizations/958457776463",
            "state": "ACTIVE",
            "updateTime": "2021-06-03T16:44:28.216Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622738668216,
      "displayName": "example-folder-level-0",
      "etag": "2f8zjBmyEStHXN45HiqtvA==",
      "lifecycleState": "ACTIVE",
      "name": "folders/299817291758",
      "parent": "organizations/958457776463",
      "updatedOn": 1622738668216,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/1065006483234",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-06-03T16:45:03.939Z",
            "displayName": "example-folder-level-1",
            "etag": "dV+Si+qLs1/yoqzfIcKDFw==",
            "name": "folders/1065006483234",
            "parent": "folders/299817291758",
            "state": "ACTIVE",
            "updateTime": "2021-06-03T16:45:03.939Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622738703939,
      "displayName": "example-folder-level-1",
      "etag": "dV+Si+qLs1/yoqzfIcKDFw==",
      "lifecycleState": "ACTIVE",
      "name": "folders/1065006483234",
      "parent": "folders/299817291758",
      "updatedOn": 1622738703939,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/811983343328",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-06-03T17:19:34.539Z",
            "displayName": "example-folder-level-2",
            "etag": "ZKljr9WBPqy9lzArWkk7Mg==",
            "name": "folders/811983343328",
            "parent": "folders/1065006483234",
            "state": "ACTIVE",
            "updateTime": "2021-06-03T17:19:34.539Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622740774539,
      "displayName": "example-folder-level-2",
      "etag": "ZKljr9WBPqy9lzArWkk7Mg==",
      "lifecycleState": "ACTIVE",
      "name": "folders/811983343328",
      "parent": "folders/1065006483234",
      "updatedOn": 1622740774539,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/306329820365",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-31T15:15:04.386Z",
            "displayName": "integration-dev",
            "etag": "AIbp3/+xtzAFwc+jklXkEA==",
            "name": "folders/306329820365",
            "parent": "organizations/958457776463",
            "state": "ACTIVE",
            "updateTime": "2021-05-31T15:15:04.386Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622474104386,
      "displayName": "integration-dev",
      "etag": "AIbp3/+xtzAFwc+jklXkEA==",
      "lifecycleState": "ACTIVE",
      "name": "folders/306329820365",
      "parent": "organizations/958457776463",
      "updatedOn": 1622474104386,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "organizations/958457776463",
      "_key": "organizations/958457776463|has|folders/299817291758",
      "_toEntityKey": "folders/299817291758",
      "_type": "google_cloud_organization_has_folder",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "folders/299817291758",
      "_key": "folders/299817291758|has|folders/1065006483234",
      "_toEntityKey": "folders/1065006483234",
      "_type": "google_cloud_folder_has_folder",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "folders/1065006483234",
      "_key": "folders/1065006483234|has|folders/811983343328",
      "_toEntityKey": "folders/811983343328",
      "_type": "google_cloud_folder_has_folder",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "organizations/958457776463",
      "_key": "organizations/958457776463|has|folders/306329820365",
      "_toEntityKey": "folders/306329820365",
      "_type": "google_cloud_organization_has_folder",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/151807220743",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/151807220743",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1632421288495,
          "displayName": "j1-gc-integration-dev-rw",
          "id": "j1-gc-integration-dev-rw",
          "lifecycleState": "DELETE_REQUESTED",
          "name": "projects/151807220743",
          "parent": "organizations/958457776463",
          "projectId": "j1-gc-integration-dev-rw",
          "updatedOn": 1632421410163,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-rw",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/538466736102",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/538466736102",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1627482617623,
          "displayName": "vmware-account",
          "id": "vmware-account",
          "lifecycleState": "ACTIVE",
          "name": "projects/538466736102",
          "parent": "organizations/958457776463",
          "projectId": "vmware-account",
          "updatedOn": 1627482619549,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=vmware-account",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/590086701852",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/590086701852",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1631743907181,
          "displayName": "Rick Sandbox",
          "id": "rick-sandbox-326122",
          "lifecycleState": "DELETE_REQUESTED",
          "name": "projects/590086701852",
          "parent": "organizations/958457776463",
          "projectId": "rick-sandbox-326122",
          "updatedOn": 1632155624084,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=rick-sandbox-326122",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/592147458023",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/592147458023",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1622559316913,
          "displayName": "My Project 36133",
          "id": "intense-pointer-315514",
          "lifecycleState": "ACTIVE",
          "name": "projects/592147458023",
          "parent": "organizations/958457776463",
          "projectId": "intense-pointer-315514",
          "updatedOn": 1628092716049,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=intense-pointer-315514",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/619127027446",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/619127027446",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1622475906340,
          "displayName": "My First Project",
          "id": "winged-guild-315415",
          "lifecycleState": "ACTIVE",
          "name": "projects/619127027446",
          "parent": "organizations/958457776463",
          "projectId": "winged-guild-315415",
          "updatedOn": 1622475908564,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=winged-guild-315415",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/723846238347",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/723846238347",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1632420695416,
          "displayName": "j1-gc-integration-dev-rick",
          "id": "j1-gc-integration-dev-rick",
          "lifecycleState": "DELETE_REQUESTED",
          "name": "projects/723846238347",
          "parent": "organizations/958457776463",
          "projectId": "j1-gc-integration-dev-rick",
          "updatedOn": 1632421052340,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-rick",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/768118999942",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/768118999942",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1626123157186,
          "displayName": "My First Project",
          "id": "glossy-calculus-319620",
          "lifecycleState": "ACTIVE",
          "name": "projects/768118999942",
          "parent": "organizations/958457776463",
          "projectId": "glossy-calculus-319620",
          "updatedOn": 1626123159133,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=glossy-calculus-319620",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/803723915367",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/803723915367",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1632155534583,
          "displayName": "jupiter-dev",
          "id": "smiling-garden-326616",
          "lifecycleState": "DELETE_REQUESTED",
          "name": "projects/803723915367",
          "parent": "organizations/958457776463",
          "projectId": "smiling-garden-326616",
          "updatedOn": 1632155613381,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=smiling-garden-326616",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/935349827748",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/935349827748",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1632155369222,
          "displayName": "jupiter-dev",
          "id": "jupiter-dev-326616",
          "lifecycleState": "ACTIVE",
          "name": "projects/935349827748",
          "parent": "organizations/958457776463",
          "projectId": "jupiter-dev-326616",
          "updatedOn": 1632155370819,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=jupiter-dev-326616",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "organizations/958457776463|has|projects/988862309073",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "organizations/958457776463",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/988862309073",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1622477706879,
          "displayName": "My First Project",
          "id": "learned-hour-315416",
          "lifecycleState": "ACTIVE",
          "name": "projects/988862309073",
          "parent": "organizations/958457776463",
          "projectId": "learned-hour-315416",
          "updatedOn": 1622477708998,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=learned-hour-315416",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_organization_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "folders/811983343328|has|projects/270453338007",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "folders/811983343328",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/270453338007",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1624364968304,
          "displayName": "j1-gc-integration-dev-nested",
          "id": "j1-gc-integration-dev-nested",
          "lifecycleState": "ACTIVE",
          "name": "projects/270453338007",
          "parent": "folders/811983343328",
          "projectId": "j1-gc-integration-dev-nested",
          "updatedOn": 1624364970108,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-nested",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_folder_has_project",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "folders/306329820365|has|projects/167984947943",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "folders/306329820365",
        "targetEntity": Object {
          "_class": Array [
            "Account",
          ],
          "_key": "projects/167984947943",
          "_rawData": undefined,
          "_type": "google_cloud_project",
          "createdOn": 1622474156125,
          "displayName": "j1-gc-integration-dev-v3",
          "id": "j1-gc-integration-dev-v3",
          "lifecycleState": "ACTIVE",
          "name": "projects/167984947943",
          "parent": "folders/306329820365",
          "projectId": "j1-gc-integration-dev-v3",
          "updatedOn": 1622474158148,
          "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_folder_has_project",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "google_cloud_project",
    "google_cloud_organization",
    "google_cloud_folder",
    "google_cloud_organization_has_folder",
    "google_cloud_folder_has_folder",
    "google_organization_has_project",
    "google_folder_has_project",
  ],
  "numCollectedEntities": 6,
  "numCollectedRelationships": 16,
}
`;

exports[`#fetchResourceManagerFolders should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Organization",
      ],
      "_key": "organizations/958457776463",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-26T21:15:56.407Z",
            "directoryCustomerId": "C048tgq5f",
            "displayName": "jupiterone.dev",
            "etag": "ZmYW46DY9wJtowGQ20pa3g==",
            "name": "organizations/958457776463",
            "state": "ACTIVE",
            "updateTime": "2021-05-26T21:15:57.254Z",
          },
        },
      ],
      "_type": "google_cloud_organization",
      "createdOn": 1622063756407,
      "directoryCustomerId": "C048tgq5f",
      "displayName": "jupiterone.dev",
      "etag": "ZmYW46DY9wJtowGQ20pa3g==",
      "lifecycleState": "ACTIVE",
      "name": "organizations/958457776463",
      "updatedOn": 1622063757254,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/299817291758",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-06-03T16:44:28.216Z",
            "displayName": "example-folder-level-0",
            "etag": "2f8zjBmyEStHXN45HiqtvA==",
            "name": "folders/299817291758",
            "parent": "organizations/958457776463",
            "state": "ACTIVE",
            "updateTime": "2021-06-03T16:44:28.216Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622738668216,
      "displayName": "example-folder-level-0",
      "etag": "2f8zjBmyEStHXN45HiqtvA==",
      "lifecycleState": "ACTIVE",
      "name": "folders/299817291758",
      "parent": "organizations/958457776463",
      "updatedOn": 1622738668216,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/1065006483234",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-06-03T16:45:03.939Z",
            "displayName": "example-folder-level-1",
            "etag": "dV+Si+qLs1/yoqzfIcKDFw==",
            "name": "folders/1065006483234",
            "parent": "folders/299817291758",
            "state": "ACTIVE",
            "updateTime": "2021-06-03T16:45:03.939Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622738703939,
      "displayName": "example-folder-level-1",
      "etag": "dV+Si+qLs1/yoqzfIcKDFw==",
      "lifecycleState": "ACTIVE",
      "name": "folders/1065006483234",
      "parent": "folders/299817291758",
      "updatedOn": 1622738703939,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/811983343328",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-06-03T17:19:34.539Z",
            "displayName": "example-folder-level-2",
            "etag": "ZKljr9WBPqy9lzArWkk7Mg==",
            "name": "folders/811983343328",
            "parent": "folders/1065006483234",
            "state": "ACTIVE",
            "updateTime": "2021-06-03T17:19:34.539Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622740774539,
      "displayName": "example-folder-level-2",
      "etag": "ZKljr9WBPqy9lzArWkk7Mg==",
      "lifecycleState": "ACTIVE",
      "name": "folders/811983343328",
      "parent": "folders/1065006483234",
      "updatedOn": 1622740774539,
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "folders/306329820365",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-31T15:15:04.386Z",
            "displayName": "integration-dev",
            "etag": "AIbp3/+xtzAFwc+jklXkEA==",
            "name": "folders/306329820365",
            "parent": "organizations/958457776463",
            "state": "ACTIVE",
            "updateTime": "2021-05-31T15:15:04.386Z",
          },
        },
      ],
      "_type": "google_cloud_folder",
      "createdOn": 1622474104386,
      "displayName": "integration-dev",
      "etag": "AIbp3/+xtzAFwc+jklXkEA==",
      "lifecycleState": "ACTIVE",
      "name": "folders/306329820365",
      "parent": "organizations/958457776463",
      "updatedOn": 1622474104386,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "organizations/958457776463",
      "_key": "organizations/958457776463|has|folders/299817291758",
      "_toEntityKey": "folders/299817291758",
      "_type": "google_cloud_organization_has_folder",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "folders/299817291758",
      "_key": "folders/299817291758|has|folders/1065006483234",
      "_toEntityKey": "folders/1065006483234",
      "_type": "google_cloud_folder_has_folder",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "folders/1065006483234",
      "_key": "folders/1065006483234|has|folders/811983343328",
      "_toEntityKey": "folders/811983343328",
      "_type": "google_cloud_folder_has_folder",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "organizations/958457776463",
      "_key": "organizations/958457776463|has|folders/306329820365",
      "_toEntityKey": "folders/306329820365",
      "_type": "google_cloud_organization_has_folder",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "google_cloud_organization",
    "google_cloud_folder",
    "google_cloud_organization_has_folder",
    "google_cloud_folder_has_folder",
  ],
  "numCollectedEntities": 5,
  "numCollectedRelationships": 4,
}
`;

exports[`#fetchResourceManagerOrganization should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Organization",
      ],
      "_key": "organizations/958457776463",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-26T21:15:56.407Z",
            "directoryCustomerId": "C048tgq5f",
            "displayName": "jupiterone.dev",
            "etag": "ZmYW46DY9wJtowGQ20pa3g==",
            "name": "organizations/958457776463",
            "state": "ACTIVE",
            "updateTime": "2021-05-26T21:15:57.254Z",
          },
        },
      ],
      "_type": "google_cloud_organization",
      "createdOn": 1622063756407,
      "directoryCustomerId": "C048tgq5f",
      "displayName": "jupiterone.dev",
      "etag": "ZmYW46DY9wJtowGQ20pa3g==",
      "lifecycleState": "ACTIVE",
      "name": "organizations/958457776463",
      "updatedOn": 1622063757254,
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "google_cloud_organization",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;

exports[`#fetchResourceManagerProject should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "CzUfI2XTWvMJgVGlic9LWA==",
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2021-05-31T15:15:58.148Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "updatedOn": 1622474158148,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "google_cloud_project",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;

exports[`#fetchResourceManagerProject should log & return Account entity if client.getProject() fails 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "CzUfI2XTWvMJgVGlic9LWA==",
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2021-05-31T15:15:58.148Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "updatedOn": 1622474158148,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "google_cloud_project",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
